physical/raft/fsm.go

```diff
@@ -243,6 +252,86 @@ func writeSnapshotMetaToDB(metadata *raft.SnapshotMeta, db *bolt.DB) error {
 	return nil
 }
 
+func (f *FSM) upgradeLocalNodeConfig() error {
+	// Read the local node config
+	var lnConfigBytes []byte
+	if err := f.db.View(func(tx *bolt.Tx) error {
+		value := tx.Bucket(configBucketName).Get(localNodeConfigKey)
+		if value != nil {
+			lnConfigBytes = make([]byte, len(value))
+			copy(lnConfigBytes, value)
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	var lnConfig LocalNodeConfigValue
+	// Entry is already present. Read and get the suffrage value.
+	if lnConfigBytes != nil {
+		err := proto.Unmarshal(lnConfigBytes, &lnConfig)
+		if err != nil {
+			return err
+		}
+		f.desiredSuffrage = lnConfig.DesiredSuffrage
+		return nil
+	}
+
+	//
+	// This is the upgrade case where there is no entry
+	//
+
+	// Refer to the persisted latest raft config
+	config := f.latestConfig.Load().(*ConfigurationValue)
+
+	// If there is no config, then this is a fresh node coming up. This could end up
+	// being a voter or non-voter. But by default assume that this is a voter. It
+	// will be changed if this node joins the cluster as a non-voter.
+	if config == nil {
+		lnConfig.DesiredSuffrage = f.desiredSuffrage
+		return f.persistDesiredSuffrage(&lnConfig)
+	}
+
+	// Get the last known suffrage of the node and assume that it is the desired
+	// suffrage. There is no better alternative here.
+	for _, srv := range config.Servers {
+		if srv.Id == f.localID {
+			switch srv.Suffrage {
+			case int32(raft.Nonvoter):
+				lnConfig.DesiredSuffrage = "non-voter"
+			default:
+				lnConfig.DesiredSuffrage = "voter"
+			}
+		}
+		// Bring the intent to the fsm instance.
+		f.desiredSuffrage = lnConfig.DesiredSuffrage
+	}
+
+	return f.persistDesiredSuffrage(&lnConfig)
+}
+
+// recordSuffrage is called when a node successfully joins the cluster. This
+// intent should land in the stored configuration. If the config isn't available
+// yet, we still go ahead and store the intent in the fsm. During the next
+// update to the configuration, this intent will be persisted.
+func (f *FSM) recordSuffrage(desiredSuffrage string) error {
+	var lnConfig LocalNodeConfigValue
+	lnConfig.DesiredSuffrage = desiredSuffrage
+	f.desiredSuffrage = desiredSuffrage
```

Comment:
Probably better to try to persist first, that way if an error occurs we haven't updated f.desiredSuffrage.

Reply:
Yep. Fixed.